pkg/services/datasources/service/secrets_mig.go

```diff
@@ -0,0 +1,52 @@
+package service
+
+import (
+	"context"
+
+	"github.com/grafana/grafana/pkg/services/datasources"
+)
+
+type DataSourceSecretMigrationService struct {
+	dataSourcesService datasources.DataSourceService
+}
+
+func ProvideDataSourceMigrationService(dataSourcesService datasources.DataSourceService) *DataSourceSecretMigrationService {
+	return &DataSourceSecretMigrationService{
+		dataSourcesService: dataSourcesService,
+	}
+}
+
+func (s *DataSourceSecretMigrationService) Migrate(ctx context.Context) error {
+	query := &datasources.GetAllDataSourcesQuery{}
+	err := s.dataSourcesService.GetAllDataSources(ctx, query)
+	if err != nil {
+		return err
+	}
+
+	for _, ds := range query.Result {
+		hasMigration := ds.JsonData.Get("secretMigrationComplete").MustBool()
+		if !hasMigration {
+			secureJsonData, err := s.dataSourcesService.DecryptedValues(ctx, ds)
+			if err != nil {
+				return err
+			}
+
+			ds.JsonData.Set("secretMigrationComplete", true)
+
+			// Secrets are set by the update data source function if the SecureJsonData is set in the command
+			// Secrets are deleted by the update data source function if the disableSecretsCompatibility flag is enabled
+			err = s.dataSourcesService.UpdateDataSource(ctx, &datasources.UpdateDataSourceCommand{
```

Comment:
That's correct, I thought by setting it to an empty map would be enough, I thought it would only check if it was not nil, but it also checks if it's empty, I also noticed the information being lost is only the one on the MustCols and MustBool functions.

I added a `sess.MustCols("secure_json_data")` and populated the command with the necessary fields.

Done on https://github.com/grafana/grafana/pull/50463/commits/bc1db8aa3b3b645a29237049cc5d072f67a347c5

Reply:
I'm afraid this wont work since we allow a datasource to be updated without providing all/any of the existing secrets. That's why I suggested a new flag/field on the `UpdateDataSourceCommand` named something like `ClearSecureData` and if that is true you add the `sess.MustCols("secure_json_data")`. It would be a similar field as below so that it cannot be set from the API consumer, only by Grafana.  https://github.com/grafana/grafana/blob/40dff288cd2fd3381775f07cf04c8de5da2c0ded/pkg/services/datasources/models.go#L131

Reply:
I noticed everywhere but provisioning calls `UpdateDataSource` on `datasource_service.go` which calls `fillWithSecureJSONData` so I think it should be safe to use  `sess.MustCols("secure_json_data")` since we always fill that information on the command for API calls. Provisioning will be updated to also use the datasource service instead of SQL Store.

Reply:
Good point. I think you're correct :+1: Just a bit scared we haven't had this set before, maybe for historic reasons. Lets try this and see how it goes.

Provisioning seems to already use the datasource service correctly as you pointed out a while ago: https://github.com/grafana/grafana/blob/c12d1727f01eeb036873e1e782a31ff3fc9dacde/pkg/services/provisioning/provisioning.go#L170

Reply:
Oh! I thing I was under the impression it used the SQLStore because the field it calls is `.store` but turns out it already uses the datasource service.